File: packages/activity-domain/src/journal.ts
import { type Infer, Schema, withParser } from '@ocentra-parent/schema-domain/effect';
import { ActivityEventIdSchema, ActivityEvidenceDigestSchema, ActivityTimestampSchema } from './primitives';
import {
  ActivityJournalCiphertextSchema,
  ActivityJournalEntryIdSchema,
  ActivityJournalNonceSchema,
} from './primitives';

export const ActivityJournalSchemaVersion = 1;

export const ActivityJournalCipherSchema = withParser(Schema.Literal('xchacha20poly1305'));

export const ActivityJournalLineSchema = withParser(
  Schema.Struct({
    schemaVersion: Schema.Literal(ActivityJournalSchemaVersion),
    entryId: ActivityJournalEntryIdSchema,
    writtenAt: ActivityTimestampSchema,
    eventId: ActivityEventIdSchema,
    cipher: ActivityJournalCipherSchema,
    nonce: ActivityJournalNonceSchema,
    ciphertext: ActivityJournalCiphertextSchema,
    activityDigest: ActivityEvidenceDigestSchema,
  })
);

export const ActivityJournalStatusSchema = withParser(
  Schema.Struct({
    schemaVersion: Schema.Literal(ActivityJournalSchemaVersion),
    encrypted: Schema.Boolean,
    entriesWritten: Schema.Number,
    bytesWritten: Schema.Number,
    lastEntryId: Schema.Union(ActivityJournalEntryIdSchema, Schema.Null),
  })
);

export type ActivityJournalCipher = Infer<typeof ActivityJournalCipherSchema>;
export type ActivityJournalLine = Infer<typeof ActivityJournalLineSchema>;
export type ActivityJournalStatus = Infer<typeof ActivityJournalStatusSchema>;

export const ActivityJournalCipher = {
  XChaCha20Poly1305: ActivityJournalCipherSchema.parse('xchacha20poly1305'),
} as const;

File: crates/agent-protocol/src/journal.rs
use serde::{Deserialize, Serialize};

pub const ACTIVITY_JOURNAL_SCHEMA_VERSION: u16 = crate::ACTIVITY_JOURNAL_SCHEMA_VERSION;

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub enum ActivityJournalCipher {
    #[serde(rename = "xchacha20poly1305")]
    XChaCha20Poly1305,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ActivityJournalLine {
    pub schema_version: u16,
    pub entry_id: String,
    pub segment_id: String,
    pub written_at: String,
    pub event_id: String,
    pub cipher: ActivityJournalCipher,
    pub nonce: String,
    pub ciphertext: String,
    pub activity_digest: String,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ActivityJournalStatus {
    pub schema_version: u16,
    pub encrypted: bool,
    pub entries_written: u64,
    pub bytes_written: u64,
    pub active_segment_id: String,
    pub segment_count: u64,
    pub rotation_max_bytes: u64,
    pub last_entry_id: Option<String>,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ActivityJournalRotationPolicy {
    pub max_segment_bytes: u64,
}
